### DISCUSSION TITLE
Can't return dictionary as response when values are numpy arrays

### DISCUSSION THREAD

**Discussion by kareemamrr** [fastapi/fastapi#7136] [Questions] [answered]
```Python
from backend import Model
from fastapi import FastAPI, File, Form, UploadFile

app = FastAPI()
model = Model()

@app.get("/")
def read_root():
    return {"Hello World"}

@app.post("/enroll")
def enroll_user(uploadedfile: UploadFile = File(...), username: str = Form(...)):
    model.enroll(uploadedfile.file, username)
    return "success"

@app.get("/db")
def view_db():
    return model._database
```
### Description
- I receive an uploaded audio file and a username and pass them to `model.enroll` to generate an embedding for that file and store it in the dictionary `_database` as `{username: embedding}.
- Embedding is of type numpy.ndarray and of shape (1, 512).
- The problem arises when I try and hit the `/db` endpoint to view the `_database` variable, it works when the dictionary is still empty but when it is populated it throws back this error `ValueError: [ValueError('dictionary update sequence element #0 has length 512; 2 is required'), TypeError('vars() argument must have __dict__ attribute')]`.
- The error seems to arise from `fastapi/encoders` line 139.
- I tried populating the dictionary with all sorts of sequences and it worked just fine.
- The expected return is {username: embedding}.

### Environment
- OS: Linux
- FastAPI Version: 0.61.1
- Python version: 3.8.2

**Comment by falkben** [comment]
Can you show what `Model` does?

My guess is that the numpy array cannot be converted to a dict. Are you specifying the dtype for the array to be `dict`?

**Comment by Kludex** [comment]
AFAIR, an encoder is needed here.

**Comment by kareemamrr** [comment]
@falkben here is `Model`
```Python
class Model:
    def __init__(self):
        self.threshold = 0.7
        self._database = {}
        self.users = []
        self.model = DeepSpeakerModel()
        self.model.m.load_weights("ResCNN_checkpoint_850.h5", by_name=True)

    def gen_mfcc(self, audio):
        return sample_from_mfcc(read_mfcc(audio, SAMPLE_RATE), NUM_FRAMES)

    def gen_embedding(self, mfcc):
        return self.model.m.predict(np.expand_dims(mfcc, axis=0))

    def get_score(self, embeds):
        return batch_cosine_similarity(embeds[0], embeds[1])

    def enroll(self, audio, name):
        mfcc = self.gen_mfcc(audio)
        embed = self.gen_embedding(mfcc)
        self._database[name] = embed
        self.users.append(name)

    def verify(self, audio, name):
        identity_embed = self._database[name]
        mfcc = self.gen_mfcc(audio)
        test_embed = self.gen_embedding(mfcc)
        score = self.get_score([identity_embed, test_embed])
        if score > self.threshold:
            return 1
        else:
            return 0
```
`_database` is already a dictionary, why would the numpy array be converted to a dict by itself?

**Comment by Mause** [comment, marked as answer]
It looks like fastapi is trying to fall back to converting the numpy array to a `dict` using `vars` as it doesn't know what to do with it. You are probably best off converting the array into a python list yourself before returning it.

```python
@app.get("/db")
def view_db():
    return {username: embedding.tolist() for username, embedding in model._database.items()}
```

**Comment by kareemamrr** [comment]
Noted. Thank you.

**Comment by tiangolo** [comment]
Thanks for the help everyone! ☕ 

And thanks for coming back to close the issue @kareemamrr 🍰 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
